src/util/reflow.rs

```diff
@@ -0,0 +1,241 @@
+use super::*;
+
+/// Encapsulate reflow logic for buffer manipulation.
+pub struct Reflow<'a> {
+    buf: &'a mut Buffer,
+    rng: Range,
+    txt: String,
+    lmt: usize,
+}
+
+impl<'a> Reflow<'a> {
+    /// Create a reflow instance, where buffer and range determine the target,
+    /// and the limit is the maximum length of a line, regardless of prefixes.
+    pub fn new(
+        buf: &'a mut Buffer, rng: Range, lmt: usize
+    ) -> std::result::Result<Self, Error> {
+        let txt = match buf.read(&rng) {
+            Some(t) => t,
+            None => bail!("Selection is invalid."),
+        };
+        Ok(Self { buf, rng, txt, lmt })
+    }
+
+    pub fn apply(mut self) -> std::result::Result<(), Error> {
+        let prefix = self.infer_prefix()?;
+        let jtxt = self.justify_str(&prefix);
+        self.buf.delete_range(self.rng.clone());
+        self.buf.cursor.move_to(self.rng.start());
+        self.buf.insert(jtxt);
+
+        Ok(())
+    }
+
+    fn infer_prefix(&self) -> std::result::Result<String, Error> {
+        match self.txt.split_whitespace().next() {
+        	Some(n) => if n.chars().next().unwrap().is_alphanumeric() {
+        	    Ok("".to_string())
+        	} else {
+        	    Ok(n.to_string())
+        	},
+        	None => bail!("Selection is empty."),
+        }
+    }
+
+
+    fn justify_str(&mut self, prefix: &str) -> String {
+        let txt = self.buf.read(&self.rng).unwrap();
+        let mut limit = self.lmt;
+        let mut justified = String::with_capacity(txt.len());
+        let mut pars = txt.split("\n\n").peekable();
+
+        let mut space_delims = ["".to_string(), " ".to_string(), "\n".to_string()];
```

Comment:
This part of the method implementation is clever, but hard to read/understand.

I put together a [Rust playground example](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=2ea8886680f6992c7ee366cb3d38eb80) of an implementation that uses `fold` to carry the length accumulator. It also avoids pre-building space delimiters, opting instead to use them literally where they're needed.

It's possible I've missed an edge case, but I think leaning into that kind of an approach is easier to follow, and would help with long-term maintainability. 😇

Reply:
The playground example you gave seems to have a few issues:
1. It'll add a newline regardless of whether or not the line is the first of a paragraph (so with the 3rd bug, it just has the bizzare effect of moving everything down one line)
2. With no prefix, everything's indented one space.
3. Sometimes it just doesn't reflow (with your playground ex, setting limit to 15 seems to do that - no idea why)
4. The first output branch doesn't seem to account for the space between the prefix and the word (though thats an easy fix)

I personally feel that once that's fixed, we'd end up with something about the same length and nearly as complex as what's currently in the fork. I completely agree with you about the complexity, but imo reflowing is non-trivial to implement trivially --- it might just be more efficient to leave it as it is.

Reply:
Fair enough! I can always refactor against the test suite afterwards. 👍🏼